### PR TITLE
Pass Go strings to Python as Unicode objects

### PR DESCRIPTION
**Pull request by ondra**
This PR builds on https://github.com/go-python/gopy/pull/170.

Currently, Python 3 CFFI returns Unicode, but Python 2 backends do not.

In Python 2, both byte and Unicode strings inherit from basestring and are interchangeable for most purposes. Some code might break, as Python 2 tries to encode Unicode strings to PYTHONIOENCODING while doing IO, and the specific encoding might be unable to represent the characters present in the string. Programs which ignore encoding _and_ have the need to output fancy characters suffer.

CPython 2 docstrings as Unicode are not implemented by this PR, but I do not think this is an issue.

Making the tests portable is difficult -- only the test for Unicode needs to handle IO encoding seriously. In the others I just strip the potential `u` which `repr` uses to mark Unicode strings in Python 2.

### CODE REVIEW

File: bind/gencffi.go [hunk 1]
```diff
@@ -87,8 +88,8 @@ class _cffi_helper(object):
 
     @staticmethod
     def cffi_cgopy_cnv_py2c_string(o):
-        if _PY3:
-            o = o.encode('ascii')
+        if _PY3 and isinstance(o, str) or not _PY3 and isinstance(o, unicode):
```
Comment (sbinet) [inline bind/gencffi.go]: could you put explicit parentheses?
with the mix of `and` and `or`, it's a bit hard to understand exactly what this is doing (well, of course "everybody" knows the precedence of `or` and `and`, but...)

Comment (ondra) [reply bind/gencffi.go]: Good idea -- it also avoids having to think about the weird operator chaning semantics in Python.

File: _examples/unicode/test.py [hunk 2]
```diff
@@ -7,17 +7,30 @@
 ## py2/py3 compat
 from __future__ import print_function, unicode_literals
 
+import sys
+
 import encoding
 
+# There is no portable way of outputting Unicode in Python 2/3 -- we encode
+# them manually and sidestep the builtin encoding machinery
+try: binary_stdout = sys.stdout.buffer  # Python 3
```
Comment (sbinet) [inline _examples/unicode/test.py]: cosmetics: could you align `binary_stdout` with the one from the line below ?

File: bind/gencpy.go [hunk 3]
```diff
@@ -297,10 +297,16 @@ func (g *cpyGen) gen() error {
...
-		g.pkg.pkg.Name(),
-		g.pkg.doc.Doc,
-	)
+	if g.pkg.doc.Doc != "" {
+		g.impl.Printf("module = Py_InitModule3(%[1]q, cpy_%[1]s_methods, %[2]q);\n\n",
+			g.pkg.pkg.Name(),
+			g.pkg.doc.Doc,
+		)
+	} else {
+		g.impl.Printf("module = Py_InitModule3(%[1]q, cpy_%[1]s_methods, 0);\n\n",
```
Comment (sbinet) [inline bind/gencpy.go]: `s/0/NULL/` ?

File: _examples/unicode/test.py (hunk 2 above)
Comment (ondra) [reply _examples/unicode/test.py]: I put the blocks on separate lines as sane people and PEP8 recommend.

File: bind/gencpy.go (hunk 3 above)
Comment (ondra) [reply bind/gencpy.go]: ok

**Pull request merged**